internal/app/backend/backend_service.go

```diff
@@ -72,52 +73,101 @@ func (s *backendService) FetchMatches(req *pb.FetchMatchesRequest, stream pb.Bac
-	startTime = time.Now()
+	errors := make(chan error, 2)
+	mmfCtx, cancelMmfs := context.WithCancel(stream.Context())
+	startMmfs := func() {
+		resultChan := make(chan mmfResult, len(req.GetProfiles()))
 
-	err = doFetchMatchesReceiveMmfResult(stream.Context(), s.mmfClients, req, resultChan)
-	if err != nil {
-		return err
-	}
+		err := doFetchMatchesReceiveMmfResult(mmfCtx, s.mmfClients, req, resultChan)
+		if err != nil {
+			// TODO: Log but continue case where mmfs were canceled once fully
+			// streaming.
+			errors <- err
+			return
+		}
 
-	proposals, err := doFetchMatchesValidateProposals(stream.Context(), resultChan, len(req.GetProfiles()))
-	if err != nil {
-		return err
-	}
+		proposals, err := doFetchMatchesValidateProposals(mmfCtx, resultChan, len(req.GetProfiles()))
+		if err != nil {
+			// TODO: Log but continue case where mmfs were canceled once fully
+			// streaming.
+			errors <- err
+			return
+		}
 
-	telemetry.RecordNUnitMeasurement(stream.Context(), mProposalCollectionPhase, time.Since(startTime).Milliseconds())
-	startTime = time.Now()
+	sendProposals:
+		for _, p := range proposals {
+			select {
+			case <-mmfCtx.Done():
+				logger.Warning("proposals from mmfs recieved too late to be sent to synchronizer.")
+				break sendProposals
+			default:
+			}
 
-	results, err := s.synchronizer.evaluate(stream.Context(), syncID, proposals)
-	if err != nil {
-		return err
-	}
+			err = syncStream.Send(&ipb.SynchronizeRequest{Proposal: p})
+			if err != nil {
+				errors <- fmt.Errorf("Error sending proposal to synchronizer: %w", err)
+				return
+			}
+		}
 
-	telemetry.RecordNUnitMeasurement(stream.Context(), mEvaluateProposalsPhase, time.Since(startTime).Milliseconds())
+		err = syncStream.CloseSend()
+		if err != nil {
+			errors <- fmt.Errorf("Error closing send stream of proposals to synchronizer: %w", err)
+			return
+		}
+		errors <- nil
+	}
 
-	for _, result := range results {
-		select {
-		case <-stream.Context().Done():
-			return stream.Context().Err()
-		default:
-			err = stream.Send(&pb.FetchMatchesResponse{Match: result})
-			telemetry.RecordUnitMeasurement(stream.Context(), mMatchesFetched)
+	go func() {
+		var startMmfsOnce sync.Once
+		defer func() {
+			startMmfsOnce.Do(func() {
+				errors <- fmt.Errorf("MMFS were never started.")
+			})
+		}()
+
+		for {
+			resp, err := syncStream.Recv()
+			if err == io.EOF {
+				errors <- nil
+				return
+			}
 			if err != nil {
-				logger.WithError(err).Error("failed to stream back the response")
-				return err
+				errors <- fmt.Errorf("Error receiving match from synchronizer: %w", err)
+				return
 			}
+
+			if resp.StartMmfs {
+				go startMmfsOnce.Do(startMmfs)
+			}
+
+			if resp.CancelMmfs {
+				cancelMmfs()
+			}
+
+			if resp.Match != nil {
+				err = stream.Send(&pb.FetchMatchesResponse{Match: resp.Match})
+				telemetry.RecordUnitMeasurement(stream.Context(), mMatchesFetched)
```

Comment:
`mmfCtx`

Reply:
Nope, this is sending a response from the backend.  mmfCtx is only used for calling the match functions.  It can be canceled without canceling the communication with the synchronizer.